### PR TITLE
perf(remote-config): memoize decoded UIConfig per ui_config topic

### PR DESCRIPTION
**Pull request by facumenzella**
### Checklist
- [x] If applicable, unit tests

### Motivation

Stacked on #7181. That PR makes `getOfferings` wait on `UiConfigProvider.getUiConfig()`, which reads and JSON-decodes the `ui_config` blobs from disk on every call — and it's also called on the workflow render path. Repeated cache hits (and a subsequent render) re-do the same disk read + decode. A reviewer flagged the avoidable work; this memoizes it.

### Description

`UiConfigProvider` now caches the decoded `UIConfig` keyed on the `ui_config` topic and reuses it while the topic is unchanged. Because blob refs are content-addressed, any content change moves the topic and invalidates the cache automatically — the same invalidation pattern `RemoteConfigSourceProvider.rebuildIfNeeded()` uses for its `sources` topic. `getUiConfig()` now reads the topic first (it's the cache key); on a hit it returns without touching disk.

Not a behavior change: the assembled `UIConfig` is identical, only recomputation is avoided. Benefits every caller — the getOfferings gate and the paywall render path.

> Note for `purchases-android`: Android's `getUiConfig` has the same repeated-decode today; worth a follow-up there.

<details>
<summary>AI session context</summary>

# AI Context

## Metadata

- PR: this PR (stacked on #7181)
- Branch: `facu/uiconfig-memory-cache` (base: `facu/offerings-gate-ui-config`)
- Author / human owner: facumenzella
- Agent(s): Claude Code (Fable 5), Codex (review loop via codex-companion)
- Generated: 2026-07-10
- Context document version: 1

## Goal

Eliminate the repeated `ui_config` disk-read + JSON-decode that #7181's readiness gate introduced on the hot path, by memoizing the decoded `UIConfig`.

## Initial Prompt

Reviewer on #7181: "we're calling `uiConfigProvider.getUiConfig()` from multiple places and every time we query the disk. Why not have UiConfigProvider cache the UIConfig in memory and only re-query whenever the ui_config topic changes?" — citing `RemoteConfigSourceProvider`'s cache-and-rebuild-on-topic-change pattern as precedent.

## Agent Contribution

- Added a memory cache to `UiConfigProvider`: `(topic, uiConfig)` under a `Lock`, reused while `manager.topic(.uiConfig)` is unchanged, re-decoded otherwise. Mirrors `RemoteConfigSourceProvider.rebuildIfNeeded()`.
- Reordered `getUiConfig()` to read the topic first (cache key), then assemble on a miss.
- Verified topic-equality invalidation is correct: `ConfigTopic` is `[String: ConfigItem]` (Equatable), and content-addressed blob refs guarantee a content change moves the topic.

## Key Implementation Decisions

- Decision: cache key is the `ui_config` `ConfigTopic`, compared by `Equatable` (like the source provider's `sourcesTopic`). Content changes move blob refs → topic changes → cache invalidates. No manual invalidation signal needed.
- Decision: async fill without holding the lock across `await` — check/read under the lock (sync), assemble on a miss without the lock, store under the lock. Two concurrent misses may both assemble (idempotent, last-writer-wins); not coalesced, matching the low call frequency.
- Decision: only successful assemblies are cached (never a nil result), so a transient failure retries next call.

## Files / Symbols Touched

- `Sources/Networking/UiConfigProvider.swift` — `getUiConfig()` split into cache-check + `assembleUiConfig`; new `cached`/`cacheLock`, `cachedUiConfig(for:)`, `store(_:for:)`.
- `Tests/UnitTests/Networking/UiConfigProviderTests.swift` — 2 tests: unchanged topic serves from memory (no extra merge/blob reads); topic change forces a re-decode.
- `Tests/UnitTests/Purchasing/OfferingsManagerTests.swift` — the gate's concurrency test updated for the new topic-first ordering (ui_config now blocks on its topic read before its blob reads).

## Validation

- RED → GREEN: the cache-hit test failed before the change (2 blob merges for 2 calls), passes after (1).
- `UiConfigProviderTests` 11/11, `WorkflowsConfigProviderTests` 11/11, `OfferingsManagerTests` 45/45 (67/67 combined). SwiftLint clean.
- Full `UnitTests` suite is broadly red in this local environment due to snapshot/signature-verification tests that fail identically on pristine `origin/main` (iOS 26 simulator); the classes exercising this change are green in isolation. CI is the authoritative gate.
- On-simulator log check (PaywallsTester, workflows on, live stress project): getOfferings still returns only after the config blobs land — network call pattern unchanged from #7181's verified run.

## Validation Gaps

- The decode-avoidance itself isn't visible in SDK logs (disk decode is silent); it's proven by the unit test's blob-read counts, not the on-device logs.
- Android not yet updated (same repeated-decode there).

## Review Focus

- Is `ConfigTopic` equality the right invalidation key (does every ui_config content change move the topic)?
- Async cache-fill: any correctness issue with two concurrent misses both assembling?

</details>

---

> [!NOTE]
> **Low Risk**
> Performance-only memoization with topic-equality invalidation; assembled UIConfig output is unchanged, with explicit guards against stale cache on topic races or nil topics.
> 
> **Overview**
> **`UiConfigProvider.getUiConfig()`** now memoizes the decoded **`UIConfig`** behind a lock, keyed on the current **`ui_config`** `ConfigTopic`. It reads the topic first; on a cache hit it returns without calling **`mergeItemsBlobData`**. Assembly is split into **`assembleUiConfig()`**, and a successful decode is stored only if the topic is unchanged after assembly (avoids caching under a stale key when remote config refreshes mid-decode). Nil/failed assemblies are not cached.
> 
> Unit tests cover cache hits, topic revision invalidation, nil topic after clear, and the mid-assembly topic race. **`MockRemoteConfigManager`** gains **`stubbedTopicSequence`** for ordered topic reads. **`OfferingsManager`** remote-config gate test is tightened to expect concurrent topic reads, then blob reads, before offerings delivery.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 9b7906368b17c029655cad75930b8f923659bd40. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: Sources/Networking/UiConfigProvider.swift [hunk 1]
```diff
@@ -13,15 +13,40 @@ import Foundation
...
     }
 
 #if !os(tvOS)
-    /// Assembles a ``UIConfig`` from the `ui_config` topic's parts. Returns `nil` when any part is unavailable
-    /// or fails to decode, so callers never render with a partially assembled configuration.
+    /// Assembles a ``UIConfig`` from the `ui_config` topic's parts. Returns `nil` when any part is
+    /// unavailable or fails to decode, so callers never render with a partially assembled
+    /// configuration. Successful results are memoized per `ui_config` topic; a topic change re-decodes.
     func getUiConfig() async -> UIConfig? {
+        let topic = await self.manager.topic(.uiConfig)
```
Comment (ajpallares) [inline Sources/Networking/UiConfigProvider.swift:32]: What's the benefit of going through the `manager` instead of going through the `diskCache` directly? https://github.com/RevenueCat/purchases-ios/blob/96fb341b99b612a36b59e01c0918ab70e3edaa48/Sources/Networking/RemoteConfigDiskCache.swift#L145-L149

This call is synchronous, so no need to `await` (but it only considers the cached uiConfig, regardless of whether there's a current call to the config endpoint).

File: Sources/Networking/UiConfigProvider.swift [hunk 2]
```diff
@@ -13,15 +13,40 @@ import Foundation
...
+        guard let uiConfig = await self.assembleUiConfig() else {
+            return nil
+        }
+        // `mergeItemsBlobData` reads the manager's live topic, so a refresh mid-assembly could
+        // decode a topic other than the key captured above. Only memoize when the topic is
+        // unchanged across the assembly.
+        if await self.manager.topic(.uiConfig) == topic {
+            self.store(uiConfig, for: topic)
+        }
+        return uiConfig
```
Comment (ajpallares) [inline Sources/Networking/UiConfigProvider.swift:46]: Should we add a test for this?

File: Sources/Networking/UiConfigProvider.swift
Comment (ajpallares) [inline Sources/Networking/UiConfigProvider.swift:1]: Since the provider doesn't observe `clearCache()` directly, correctness after an identity change relies entirely on `topic(.uiConfig)` going to `nil` and `cachedUiConfig(for: nil)` therefore missing. That works, but it's implicit. Could we add a test that sets the cache, then makes the topic `nil`, and asserts we don't serve the stale value? (The stale tuple also just lingers in memory until overwritten; probably fine, but a test would make the intent explicit.)

File: Sources/Networking/UiConfigProvider.swift [hunk 3]
```diff
@@ -13,15 +13,40 @@ import Foundation
 final class UiConfigProvider {
 
     private let manager: RemoteConfigManagerType
+    private let cacheLock = Lock()
+    // Last decoded config and the `ui_config` topic it was built from. Reused while the topic
+    // is unchanged so repeated callers (the offerings gate, workflow render) don't re-read and
+    // re-decode the blobs on every call. Blob refs are content-addressed, so any content change
+    // moves the topic and invalidates this.
+    private var cached: (topic: RemoteConfiguration.ConfigTopic, uiConfig: UIConfig)?
```
Comment (tonidero) [inline Sources/Networking/UiConfigProvider.swift:21]: Just wondering, have you benchmarked how much time is spent decoding UIConfig (in total, if done multiple times)? Asking because ideally, we keep the blobs out of memory, to avoid having huge memory usage (one of the intentions of the config endpoint). 
If this is called a lot, I think definitely we should optimizer it as you did though :)

File: Sources/Networking/UiConfigProvider.swift (hunk 2 above)
Comment (facumenzella) [reply Sources/Networking/UiConfigProvider.swift:46]: Added `testDoesNotCacheWhenTopicChangesDuringAssembly`: a mock topic sequence returns topic A for the cache-key read and B for the post-assembly re-read, and the test asserts a later call re-decodes (nothing was memoized under A). 83fdb1b20

File: Sources/Networking/UiConfigProvider.swift
Comment (facumenzella) [reply Sources/Networking/UiConfigProvider.swift:1]: Made it explicit with `testDoesNotServeStaleUiConfigAfterTopicBecomesNil`: it caches, then sets the committed topic to nil (and clears the blobs) and asserts `getUiConfig()` returns nil rather than the stale cached value. You're right that the lingering tuple is harmless (it's only ever returned on an exact topic-equality hit, which a nil topic can't produce); the test pins that intent. 83fdb1b20

**Review by tonidero — COMMENTED** [review]
Nice improvement! Just a question other than Antonio's comments.

File: Sources/Networking/UiConfigProvider.swift (hunk 1 above)
Comment (facumenzella) [reply Sources/Networking/UiConfigProvider.swift:32]: Good question. `UiConfigProvider` only holds the `manager`, and `RemoteConfigManagerType` exposes topics solely through `topic(_:) async` — the manager is actor-isolated and its `diskCache` is private, so there's no synchronous committed-topic read reachable from here. Going through `diskCache.topic` directly would mean injecting the disk cache (or a topic store) into `UiConfigProvider`, which neither construction site (`OfferingsManager`, `WorkflowsConfigProvider`) currently has — a wider DI change than this perf PR.

On the `await`/refresh concern: it's benign here. On a cache hit the topic is committed, so `readCommittedState(refreshIfMissing:)` returns immediately without a refresh; `refreshIfMissing` could only fire when the topic is absent, in which case nothing was cached anyway. So the extra work is one actor hop, not a refresh. Happy to do the sync-diskCache injection as a small follow-up if you'd prefer it — just didn't want to widen this PR's surface without a nudge.

File: Tests/UnitTests/Purchasing/OfferingsManagerTests.swift [hunk 4]
```diff
@@ -1116,11 +1117,16 @@ extension OfferingsManagerTests {
         let delivered: Atomic<Bool> = false
         manager.offerings(appUserID: MockData.anyAppUserID) { _ in delivered.value = true }
 
+        // Both readiness branches block on their topic read.
         expect(mockRemoteConfigManager.invokedTopicCount).toEventually(beGreaterThan(0))
```
Comment (ajpallares) [inline Tests/UnitTests/Purchasing/OfferingsManagerTests.swift]: Should we assert here that both awaits happen concurrently? It's an implementation detail, but I think it's important for get offerings to be as quick as possible
```suggestion
        expect(mockRemoteConfigManager.invokedTopicCount).toEventually(beGreaterThanOrEqualTo(2))
```

File: Tests/UnitTests/Networking/UiConfigProviderTests.swift [hunk 5]
```diff
@@ -142,6 +142,90 @@ class UiConfigProviderTests: TestCase {
...
+        self.mockManager.stubbedTopicSequence = [topicA, topicB]
+
+        _ = await self.provider.getUiConfig()
+        expect(self.mockManager.invokedMergeItemsBlobDataParameters.count) == 1
+
+        // Sequence exhausted; topic() now returns the stubbed topic (A). Nothing was cached
+        // under A, so this re-decodes rather than serving a value assembled during the race.
+        self.mockManager.stubbedTopics[.uiConfig] = topicA
+        _ = await self.provider.getUiConfig()
+        expect(self.mockManager.invokedMergeItemsBlobDataParameters.count) == 2
```
Comment (ajpallares) [inline Tests/UnitTests/Networking/UiConfigProviderTests.swift:187]: nit: isn't this fragile in that it relies on the mock's `mergeItemsBlobData` not calling `topic()`? The real `mergeItemsBlobData` reads the topic per blob, so if the mock ever becomes more faithful this test would consume the sequence differently and quietly mis-test. Might be worth a short comment noting that dependency so a future mock refactor doesn't break the intent unnoticed.

File: Sources/Networking/UiConfigProvider.swift (hunk 3 above)
Comment (ajpallares) [reply Sources/Networking/UiConfigProvider.swift:21]: FWIW, [I proposed this change](https://revenuecat.slack.com/archives/C0A5704AFN0/p1783671949132419?thread_ts=1783665211.592389&cid=C0A5704AFN0) because I noticed we were reading the file from multiple places (sometimes 3 times for get offerings). And I figured that `UIConfig` would be small. But Toni raises a good point. 

I was thinking of what I did for the `sources` topic, which we know is small (no blobs), but I'm not so sure it's true for `UIConfig`.

File: Sources/Networking/UiConfigProvider.swift (hunk 1 above)
Comment (facumenzella) [reply Sources/Networking/UiConfigProvider.swift:32]: `UiConfigProvider` only holds the `RemoteConfigManagerType` protocol, which doesn't expose `diskCache` (it's private on the concrete manager), so a synchronous read isn't reachable from here. Going through `manager.topic(.uiConfig)` also keeps the cache key consistent with what `mergeItemsBlobData` reads (both go through the manager's committed-state read), and on a warm topic it returns without triggering a refresh, so the hit path stays cheap.

File: Sources/Networking/UiConfigProvider.swift (hunk 2 above)
Comment (facumenzella) [reply Sources/Networking/UiConfigProvider.swift:46]: Added `testDoesNotCacheWhenTopicChangesDuringAssembly`: it stubs a topic sequence so the cache-key read sees topic A and the post-assembly re-read sees topic B, then asserts nothing was memoized under A (the next call re-decodes).

File: Tests/UnitTests/Purchasing/OfferingsManagerTests.swift (hunk 4 above)
Comment (facumenzella) [reply Tests/UnitTests/Purchasing/OfferingsManagerTests.swift]: Applied.

File: Tests/UnitTests/Networking/UiConfigProviderTests.swift (hunk 5 above)
Comment (facumenzella) [reply Tests/UnitTests/Networking/UiConfigProviderTests.swift:187]: Added a comment noting the dependency: the test relies on the mock's `mergeItemsBlobData` not calling `topic()`, since `getUiConfig` makes exactly two topic reads per call (cache-key read + post-assembly re-read).

File: Sources/Networking/UiConfigProvider.swift (hunk 3 above)
Comment (facumenzella) [reply Sources/Networking/UiConfigProvider.swift:21]: I am running a 50 iteration test with the benchmark project. Will post the results here 💅

**Comment by facumenzella** [comment]
Closing this. We benchmarked it and the cache doesn't help on the paths that matter.

Against the stress project (188 offerings / 162 workflows):
- A/B'd a 50-pass round with vs without the cache: no cold win and no warm win, both within run-to-run noise (config cold moved +7%, but the legacy control also drifted +3%, so effectively flat).
- Instrumented the readiness gate directly: the `ui_config` decode is ~4ms per call. So even the "3 reads per get offerings" I was worried about only add up to ~12ms. There just isn't much to save here.

Why the cache never fires: every launch is a fresh process (empty in-memory cache), and within a launch the gate uses a throwaway `UiConfigProvider` while the render uses a separate long-lived one, so they never share it.

Where the cold time actually goes: `/offerings` serializes behind `/config` on the shared serial backend queue and sits ~1.2s behind it. That is the thing worth fixing (same idea as purchases-android #3751), not the decode.

If we ever do want to dedupe the decode, the real fix is sharing one `UiConfigProvider` between the gate and the render (that would also make a cache like this one actually hit), but the numbers say it is not worth it right now. Thanks for the idea though, it was worth checking.
